File: web/src/services/logs.ts
import api, { ApiResponse } from './api';

// Access log interfaces
export interface AccessLog {
  id: number;
  link_id: number;
  accessed_at: string;
  ip_address: string;
  user_agent: string;
  referer?: string;
  country?: string;
  city?: string;
  device_type?: string;
  browser?: string;
  os?: string;
}

// Log service class
class LogService {  /**
   * Get access logs with optional filtering
   * @param linkId - Optional link ID to filter logs by
   * @param limit - Optional limit for number of logs to return
   * @returns Promise with array of access logs
   */
  async getLogs(linkId?: string, limit?: number): Promise<AccessLog[]> {
    try {
      const params: Record<string, string> = {};
      if (linkId) params.link_id = linkId;
      if (limit) params.limit = limit.toString();

      const response = await api.get<{ logs: AccessLog[], message: string, count: number }>('/logs', { params });
      return response.data.logs || [];
    } catch (error: any) {
      const errorMessage = error.response?.data?.error || 'Failed to fetch access logs';
      throw new Error(errorMessage);
    }
  }

  /**
   * Get access logs for a specific user
   * @param userID - User ID to fetch logs for
   * @param limit - Optional limit for number of logs to return
   * @returns Promise with array of access logs
   */
  async getLogsByUser(userID: string, limit?: number): Promise<AccessLog[]> {
    try {
      const params: Record<string, string> = { user_id: userID };
      if (limit) params.limit = limit.toString();

      const response = await api.get<{ logs: AccessLog[], message: string, count: number }>('/logs/user', { params });
      return response.data.logs || [];
    } catch (error: any) {
      const errorMessage = error.response?.data?.error || 'Failed to fetch user access logs';
      throw new Error(errorMessage);
    }
  }
}

// Create and export a singleton instance
const logService = new LogService();
export default logService;

File: internal/handlers/logs/logs.go
package logs

import (
	"encoding/json"
	"fmt"
	"link-guardian/internal/models"
	"link-guardian/internal/repositories/db"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

// geoIPResult is the structure for ip-api.com response
type geoIPResult struct {
	Country string `json:"countryCode"`
	City    string `json:"city"`
}

// getGeoIPInfo fetches country and city from ip-api.com
func getGeoIPInfo(ip string) (country, city string) {
	url := fmt.Sprintf("http://ip-api.com/json/%s?fields=countryCode,city", ip)
	client := http.Client{Timeout: 2 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return "", ""
	}
	defer resp.Body.Close()
	var result geoIPResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", ""
	}
	return result.Country, result.City
}

// LogLinkAccess logs a link access event to the database
func LogLinkAccess(linkID int, ipAddress, userAgent, referer string, accessedAt time.Time) error {
	// Parse deviceType, browser, os from userAgent
	deviceType, browser, os := db.ParseUserAgent(userAgent)
	country, city := getGeoIPInfo(ipAddress)
	fmt.Println("Country:", country, "City:", city)
	return db.InsertAccessLogToDB(linkID, ipAddress, userAgent, referer, country, city, deviceType, browser, os, accessedAt)
}

// ListAccessLogsHandler returns access logs, optionally filtered by link_id and limited
func ListAccessLogsHandler(c *gin.Context) {
	linkID := c.Query("link_id")
	limitStr := c.Query("limit")
	limit := 50
	if limitStr != "" {
		if l, err := strconv.Atoi(limitStr); err == nil && l > 0 && l <= 100 {
			limit = l
		}
	}

	logs, err := db.GetAccessLogs(linkID, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch access logs"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"logs": logs})
}

// ListAccessLogsByUserHandler returns access logs for a specific user
func ListAccessLogsByUserHandler(c *gin.Context) {
	userID := c.Query("user_id")
	username := c.Query("username")
	limitStr := c.Query("limit")
	limit := 50
	if limitStr != "" {
		if l, err := strconv.Atoi(limitStr); err == nil && l > 0 && l <= 100 {
			limit = l
		}
	}

	var logs []models.AccessLog
	var err error

	// Check if we have a user identifier
	if userID != "" {
		// Convert user ID to int
		uid, err := strconv.Atoi(userID)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID format"})
			return
		}
		logs, err = db.GetAccessLogsByUser(uid, limit)
	} else if username != "" {
		logs, err = db.GetAccessLogsByUser(username, limit)
	} else {
		// Fallback to regular access logs if no user identifier is provided
		linkID := c.Query("link_id")
		logs, err = db.GetAccessLogs(linkID, limit)
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch access logs"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"logs": logs})
}
